{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "561c1f00-f7f5-4a50-80d5-795d7aabdf9f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import multiprocessing\n",
    "import subprocess\n",
    "import random\n",
    "import pickle\n",
    "import shutil\n",
    "import glob\n",
    "import math\n",
    "import csv\n",
    "import sys\n",
    "import os\n",
    "import re\n",
    "\n",
    "from pybedtools import BedTool\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib import animation\n",
    "import seaborn as sns\n",
    "from matplotlib_venn import venn2\n",
    "from matplotlib_venn import venn3\n",
    "plt.rcParams['figure.figsize'] = (20.0, 10.0)\n",
    "plt.rcParams['font.family'] = \"serif\"\n",
    "plt.rcParams['font.size'] = 24\n",
    "%matplotlib inline\n",
    "\n",
    "gff3Cols=[\"seqid\",\"source\",\"type\",\"start\",\"end\",\"score\",\"strand\",\"phase\",\"attributes\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aae803d5-9da3-4c02-866b-5c8c3502d9fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "base_dir = \"./\"\n",
    "out_dir = base_dir+\"step3/\"\n",
    "soft_dir = base_dir+\"soft/\"\n",
    "\n",
    "chess3_gtf_fname = base_dir+\"step2/chess.gtf\"\n",
    "\n",
    "data_dir = base_dir+\"data/\"\n",
    "gencode_gtf_fname = data_dir+\"latest_gtfs/gencode.primary.gtf\"\n",
    "refseq_gtf_fname = data_dir+\"latest_gtfs/refseq.primary.gtf\"\n",
    "mane_gtf_fname = data_dir+\"latest_gtfs/MANE.primary.gtf\"\n",
    "refgen_gtf_fname = data_dir+\"latest_gtfs/refgen.primary.gtf\"\n",
    "chess2_gtf_fname = data_dir+\"latest_gtfs/chess2.gtf\"\n",
    "M1_gtf_fname = data_dir+\"latest_gtfs/1M.primary.gtf\"\n",
    "M27_gtf_fname = data_dir+\"latest_gtfs/27M.primary.gtf\"\n",
    "\n",
    "seeder_fname = soft_dir+\"scripts/genomic_scripts/seeder/seeder\"\n",
    "reader_fname = soft_dir+\"scripts/genomic_scripts/reader/reader\"\n",
    "adder_fname = soft_dir+\"scripts/genomic_scripts/adder/adder\"\n",
    "orfanage_fname = soft_dir+\"scripts/genomic_scripts/orfanage/orfanage\"\n",
    "ender_fname = soft_dir+\"scripts/genomic_scripts/ender/ender\"\n",
    "ppp_bin = \"phylocsf++\"\n",
    "\n",
    "ref_fasta_fname = \"hg38_p12_ucsc.no_alts.no_fixs.fa\"\n",
    "ppp_track = \"tracks/PhyloCSF+1.bw\"\n",
    "\n",
    "num_threads = 40"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "642c0a8d-184e-4361-b266-cd01f9352843",
   "metadata": {},
   "outputs": [],
   "source": [
    "if not os.path.exists(out_dir):\n",
    "    os.makedirs(out_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "155d182a-dfaa-43e0-8dc1-a6e527e89777",
   "metadata": {},
   "outputs": [],
   "source": [
    "# since gffcompare does not readily recognize single-exon transcripts we need to use a dedicated solution\n",
    "# use coverage data from tiebrush to check whether a single-exon transcript is covered\n",
    "# since alignment of single-exon transcripts does not really care for the strand - we disregard strandedness\n",
    "# anything with coverage >= % is included\n",
    "\n",
    "out_gtf_fname = M27_gtf_fname.rstrip(\".gtf\")+\".seeder.gtf\"\n",
    "seeder_cmd = [seeder_fname,\n",
    "             \"-i\",M27_gtf_fname,\n",
    "             \"-r\",mane_gtf_fname,\n",
    "             \"-o\",out_gtf_fname,\n",
    "             \"-p\",\"80\"]\n",
    "print(\" \".join(seeder_cmd))\n",
    "subprocess.call(seeder_cmd)\n",
    "\n",
    "# load all IDs already in chess\n",
    "chess_tids = set()\n",
    "\n",
    "with open(chess3_gtf_fname,\"r\") as inFP:\n",
    "    for line in inFP:\n",
    "        lcs = line.rstrip(\"\\n\").split(\"\\t\")\n",
    "        if lcs[2]==\"transcript\":\n",
    "            tid = lcs[8].split(\"transcript_id \\\"\",1)[1].split(\"\\\"\",1)[0]\n",
    "            chess_tids.add(tid)\n",
    "\n",
    "# add missing transcripts to chess\n",
    "seeder_out_gtf_fname = out_dir+chess3_gtf_fname.split(\"/\")[-1].rstrip(\".gtf\")+\".seeder.gtf\"\n",
    "\n",
    "with open(seeder_out_gtf_fname,\"w+\") as outFP:\n",
    "    with open(chess3_gtf_fname,\"r\") as inFP: # write previous version first\n",
    "        for line in inFP:\n",
    "            outFP.write(line)\n",
    "    with open(out_gtf_fname,\"r\") as inFP:\n",
    "        for line in inFP:\n",
    "            lcs = line.rstrip(\"\\n\").split(\"\\t\")\n",
    "            if lcs[2]==\"transcript\":\n",
    "                tid = lcs[8].split(\"transcript_id \\\"\",1)[1].split(\"\\\"\",1)[0]\n",
    "                seeder_type = lcs[8].split(\"seeder_match \\\"\",1)[1].split(\"\\\"\",1)[0]\n",
    "                if tid in chess_tids or seeder_type == \"0\":\n",
    "                    write_cur = False\n",
    "                    continue\n",
    "                else:\n",
    "                    write_cur = True\n",
    "            \n",
    "            if write_cur:\n",
    "                if lcs[2]==\"exon\" or lcs[2]==\"transcript\": # skip cds for now if present\n",
    "                    outFP.write(line)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b52d9f91-573e-47fa-8174-4aa7e0be7084",
   "metadata": {},
   "outputs": [],
   "source": [
    "# run orfanage and ppp to append CDSs\n",
    "chess_gtf_fname = seeder_out_gtf_fname\n",
    "orfppp_out_gtf_fname = chess_gtf_fname.rstrip(\".gtf\")+\".cds.gtf\"\n",
    "orfppp_cmd = [base_dir+\"soft/pipeline/run_orfanage_ppp.py\",\n",
    "             \"--input\",chess_gtf_fname,\n",
    "             \"--phylocsfpp\",ppp_bin,\n",
    "             \"--threads\",str(num_threads),\n",
    "             \"--ppp_idx\",ppp_track,\n",
    "             \"--reference\",ref_fasta_fname,\n",
    "             \"--orfanage\",orfanage_fname,\n",
    "             \"--output\",orfppp_out_gtf_fname,\n",
    "             \"--annotations\",gencode_gtf_fname+\",\"+refseq_gtf_fname+\",\"+mane_gtf_fname]\n",
    "print(\" \".join(orfppp_cmd))\n",
    "# subprocess.call(orfppp_cmd)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cba8ed22-d820-4d7d-904f-1f32f48b4b5b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# search for CRS\n",
    "chess_gtf_fname = orfppp_out_gtf_fname\n",
    "crs_out_gtf_fname = chess_gtf_fname.rstrip(\".gtf\")+\".crs.gtf\"\n",
    "crs_cmd = [base_dir+\"soft/pipeline/find_crs.py\",\n",
    "             \"--input\",chess_gtf_fname,\n",
    "             \"--output\",crs_out_gtf_fname,\n",
    "             \"--crs\",base_dir+\"data/crfs/Supplemental_Data_1__patch_2020-07-28_fix_hg38_coordinates.csv\"]\n",
    "print(\" \".join(crs_cmd))\n",
    "subprocess.call(crs_cmd)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3576638f-4e67-4771-a4fc-7b28eb99640f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# examine short exons\n",
    "chess_gtf_fname = crs_out_gtf_fname\n",
    "short_out_gtf_fname = chess_gtf_fname.rstrip(\".gtf\")+\".short_label.gtf\"\n",
    "short_cmd = [base_dir+\"soft/pipeline/examine_short_exons.py\",\n",
    "             \"--input\",chess_gtf_fname,\n",
    "             \"--output\",short_out_gtf_fname,\n",
    "             \"--annotations\",\"GENCODE:\"+gencode_gtf_fname+\",RefSeq:\"+refseq_gtf_fname+\",MANE:\"+mane_gtf_fname]\n",
    "print(\" \".join(short_cmd))\n",
    "subprocess.call(short_cmd)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b7714030-d2b9-447d-aa71-124b4f30b8b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# fix ends of transcripts\n",
    "chess_gtf_fname = short_out_gtf_fname\n",
    "ender_out_gtf_fname = chess_gtf_fname.rstrip(\".gtf\")+\".ender.gtf\"\n",
    "ender_cmd = [ender_fname,\n",
    "             \"-s\",\"-t\",\"100\",\"-m\",\"100\",\n",
    "             \"-i\",chess_gtf_fname,\n",
    "             \"-r\",mane_gtf_fname+\",\"+refseq_gtf_fname+\",\"+gencode_gtf_fname,\n",
    "             \"-o\",ender_out_gtf_fname]\n",
    "\n",
    "print(\" \".join(ender_cmd))\n",
    "subprocess.call(ender_cmd)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "14fdcf66-6c10-4ce5-a869-beaa290a9b21",
   "metadata": {},
   "outputs": [],
   "source": [
    "# replace chromosome M\n",
    "chess_gtf_fname = ender_out_gtf_fname\n",
    "chrM_out_gtf_fname = chess_gtf_fname.rstrip(\".gtf\")+\".chrM.gtf\"\n",
    "chrM_cmd = [base_dir+\"soft/pipeline/replace_chrM.py\",\n",
    "             \"--input\",chess_gtf_fname,\n",
    "             \"--output\",chrM_out_gtf_fname,\n",
    "             \"--annotation\",gencode_gtf_fname]\n",
    "print(\" \".join(chrM_cmd))\n",
    "subprocess.call(chrM_cmd)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "03ff3db3-4695-4be7-a946-cd45a618ef02",
   "metadata": {},
   "outputs": [],
   "source": [
    "# add mising transcripts from MANE\n",
    "chess_gtf_fname = chrM_out_gtf_fname\n",
    "adder_out_gtf_fname = chess_gtf_fname.rstrip(\".gtf\")+\".adder.gtf\"\n",
    "adder_cmd = [adder_fname,\n",
    "             \"-i\",chess_gtf_fname,\n",
    "             \"-r\",\"MANE:\"+mane_gtf_fname,\n",
    "             \"-o\",adder_out_gtf_fname]\n",
    "\n",
    "print(\" \".join(adder_cmd))\n",
    "subprocess.call(adder_cmd)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "df7c8dc7-490f-4cd7-a218-68a0eba52bd4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# filter readthroughs\n",
    "chess_gtf_fname = adder_out_gtf_fname\n",
    "reader_out_gtf_fname = chess_gtf_fname.rstrip(\".gtf\")+\".reader.gtf\"\n",
    "reader_cmd = [base_dir+\"soft/pipeline/filter_readthrough.py\",\n",
    "             \"--input\",chess_gtf_fname,\n",
    "             \"--annotations\",\"RefGEN:\"+refgen_gtf_fname+\",MANE:\"+mane_gtf_fname+\",RefSeq:\"+refseq_gtf_fname,\n",
    "             \"--output\",reader_out_gtf_fname,\n",
    "             \"--reader\",reader_fname]\n",
    "\n",
    "print(\" \".join(reader_cmd))\n",
    "subprocess.call(reader_cmd)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0393e011-592e-4e55-8ee0-25b1444a5af2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# assign gene and transcript IDs with namer\n",
    "chess_gtf_fname = reader_out_gtf_fname\n",
    "namer_out_gtf_fname = chess_gtf_fname.rstrip(\".gtf\")+\".namer.gtf\"\n",
    "namer_cmd = [namer_fname,\n",
    "             \"-i\",chess_gtf_fname,\n",
    "             \"-r\",chess2_gtf_fname,\n",
    "             \"-o\",namer_out_gtf_fname]\n",
    "\n",
    "print(\" \".join(namer_cmd))\n",
    "subprocess.call(namer_cmd)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "146fabda-b87e-4c06-85be-795a167b8692",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
