{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "2b8a3ff7-36f8-4cc0-8194-de1366c0dfa0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import multiprocessing\n",
    "import subprocess\n",
    "import random\n",
    "import pickle\n",
    "import shutil\n",
    "import glob\n",
    "import math\n",
    "import csv\n",
    "import sys\n",
    "import os\n",
    "import re\n",
    "\n",
    "from pybedtools import BedTool\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib import animation\n",
    "import seaborn as sns\n",
    "from matplotlib_venn import venn2\n",
    "from matplotlib_venn import venn3\n",
    "plt.rcParams['figure.figsize'] = (20.0, 10.0)\n",
    "plt.rcParams['font.family'] = \"serif\"\n",
    "plt.rcParams['font.size'] = 24\n",
    "%matplotlib inline\n",
    "\n",
    "gff3Cols=[\"seqid\",\"source\",\"type\",\"start\",\"end\",\"score\",\"strand\",\"phase\",\"attributes\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "80d23dac-d521-4040-8cc8-a8eea114cf2e",
   "metadata": {},
   "outputs": [],
   "source": [
    "base_dir = \"./\"\n",
    "out_dir = base_dir+\"step6/\"\n",
    "soft_dir = base_dir+\"soft/\"\n",
    "\n",
    "chess3_gtf_fname = bae_dir+\"step5/chess.gtf\"\n",
    "\n",
    "data_dir = base_dir+\"data/\"\n",
    "gencode_gtf_fname = data_dir+\"latest_gtfs/gencode.primary.gtf\"\n",
    "refseq_gtf_fname = data_dir+\"latest_gtfs/refseq.primary.gtf\"\n",
    "mane_gtf_fname = data_dir+\"latest_gtfs/MANE.primary.gtf\"\n",
    "M27_gtf_fname = data_dir+\"latest_gtfs/27M.primary.gtf\"\n",
    "chess2_gtf_fname = data_dir+\"latest_gtfs/chess2.gtf\"\n",
    "\n",
    "ref_fasta_fname = \"hg38_p12_ucsc.no_alts.no_fixs.fa\"\n",
    "remove_gtf = data_dir+\"chess_subset.gtf\"\n",
    "\n",
    "gffcompare_bin = \"gffcompare\"\n",
    "\n",
    "num_threads = 40"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "07d7625f-049f-4943-b988-ad70cab9fbbb",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_by_tag(gtf_fname,kvs,out_fp): # kvs is a dict of individual keys to sets of passing labels\n",
    "    with open(gtf_fname,\"r\") as inFP:\n",
    "        write_tx = False\n",
    "        for line in inFP:\n",
    "            lcs = line.strip().split(\"\\t\")\n",
    "            if lcs[2]==\"transcript\":\n",
    "                write_tx = False\n",
    "                atts = [(x.split(\" \\\"\",1)[0].strip(),x.split(\" \\\"\",1)[1].rstrip(\"\\\"\")) for x in lcs[8].split(\";\")]\n",
    "                for ak,av in atts:\n",
    "                    if ak in kvs:\n",
    "                        if av in kvs[ak]:\n",
    "                            write_tx=True\n",
    "                            break\n",
    "            if write_tx:\n",
    "                out_fp.write(line)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "92f189d5-38a4-4416-9f4b-e0d6d365953b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# select selenocysteine (gencode: tag \"seleno\" - 79 txs)\n",
    "\n",
    "gencode_passing_kvs = {\"tag\":{\"seleno\"}}\n",
    "out_gencode_fname = out_dir+\"must_include.gencode.gtf\"\n",
    "out_gencode_fp = open(out_gencode_fname,\"w+\")\n",
    "get_by_tag(gencode_gtf_fname,gencode_passing_kvs,out_gencode_fp)\n",
    "out_gencode_fp.close()\n",
    "\n",
    "refseq_passing_kvs = {\"tag\":{\"MANE Select\",\"RefSeq Select\"},\n",
    "                      \"gbkey\":{\"V_region\", \"V_segment\", \"D_segment\", \"J_segment\", \"C_region\",\n",
    "                               \"tRNA\", \"rRNA\", \"snRNA\", \"snoRNA\", \"tmRNA\",\"Y_RNA\"},\n",
    "                      \"gene_biotype\":{\"V_region\", \"V_segment\", \"D_segment\", \"J_segment\", \"C_region\",\n",
    "                                      \"tRNA\", \"rRNA\", \"snRNA\", \"snoRNA\", \"tmRNA\",\"Y_RNA\"}}\n",
    "out_refseq_fname = out_dir+\"must_include.refseq.gtf\"\n",
    "out_refseq_fp = open(out_refseq_fname,\"w+\")\n",
    "get_by_tag(refseq_gtf_fname,refseq_passing_kvs,out_refseq_fp)\n",
    "out_refseq_fp.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "8b5a1698-c57d-461d-b9df-56e5f084c4be",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [],
   "source": [
    "def load_gid_tid(gtf_fname): # load dictionary of gid to max_tid\n",
    "    gtf = pd.read_csv(gtf_fname,sep=\"\\t\",comment=\"#\",names=gff3Cols)\n",
    "    gtf = gtf[gtf[\"type\"]==\"transcript\"].reset_index(drop=True)\n",
    "    gtf[\"tmp\"] = gtf[\"attributes\"].str.split(\"transcript_id \\\"\",expand=True,n=1)[1].str.split(\"\\\"\",expand=True,n=1)[0]\n",
    "    gtf[\"gid\"] = gtf[\"tmp\"].str.split(\".\",expand=True,n=2)[1].astype(int)\n",
    "    gtf[\"tid\"] = gtf[\"tmp\"].str.split(\".\",expand=True,n=2)[2].astype(int)\n",
    "    gtf=gtf[[\"gid\",\"tid\"]]\n",
    "    ggtf = gtf.groupby(by=\"gid\").max().reset_index()\n",
    "    gdict = ggtf.set_index('gid').to_dict()['tid']\n",
    "    return gdict\n",
    "\n",
    "def load_tid_gid(gtf_fname): # load dictionary of tid to gid\n",
    "    gtf = pd.read_csv(gtf_fname,sep=\"\\t\",comment=\"#\",names=gff3Cols)\n",
    "    gtf = gtf[gtf[\"type\"]==\"transcript\"].reset_index(drop=True)\n",
    "    tdict = dict()\n",
    "    gtf[\"tid\"] = gtf[\"attributes\"].str.split(\"transcript_id \\\"\",expand=True,n=1)[1].str.split(\"\\\"\",expand=True,n=1)[0]\n",
    "    gtf[\"gid\"] = gtf[\"attributes\"].str.split(\"gene_id \\\"\",expand=True,n=1)[1].str.split(\"\\\"\",expand=True,n=1)[0]\n",
    "    gtf=gtf[[\"gid\",\"tid\"]]\n",
    "    tdict = gtf.set_index('tid').to_dict()['gid']\n",
    "    return tdict\n",
    "\n",
    "def load_tid_rec(gtf_fname): # load dictionary of gtf to full transcript (transcript,exons,cds) record\n",
    "    recs = dict()\n",
    "    with open(gtf_fname,\"r\") as inFP:\n",
    "        for line in inFP:\n",
    "            if len(line.strip())==0:\n",
    "                continue\n",
    "            lcs = line.strip().split(\"\\t\")\n",
    "            tid = lcs[8].split(\"transcript_id \\\"\",1)[1].split(\"\\\"\",1)[0]\n",
    "            if lcs[2]==\"transcript\":\n",
    "                assert tid not in recs,\"duplicate tids: \"+tid\n",
    "                recs[tid]=line\n",
    "            else:\n",
    "                assert tid in recs,\"tid not found: \"+tid\n",
    "                recs[tid]+=line\n",
    "                \n",
    "    return recs\n",
    "\n",
    "def add_from_tmap(ref_gtf_fname,qry_gtf_fname,tmap,out_fname,kv2add,qtmap):\n",
    "    outFP = open(out_fname,\"w+\")\n",
    "    \n",
    "    ref_gdict = load_gid_tid(ref_gtf_fname)\n",
    "    next_gid = max(ref_gdict)+1\n",
    "    \n",
    "    \n",
    "    # load qry into dict (tid to gtf record)\n",
    "    qry_recs = load_tid_rec(qry_gtf_fname)\n",
    "    qry_tdict = load_tid_gid(qry_gtf_fname)  \n",
    "    \n",
    "    # parse tmap and decide from the code what to do\n",
    "    pass_codes = set([\"=\",\"c\",\"k\",\"j\",\"m\",\"n\"])\n",
    "\n",
    "    tmp = pd.read_csv(tmap,sep=\"\\t\",usecols=[0,1,2,3,4])\n",
    "    \n",
    "    loaded_tids = set() # keeps track of which tids have been used to not overwrite with reference\n",
    "\n",
    "    for qgid,grp in tmp.groupby(by=\"qry_gene_id\"):\n",
    "        rgids = set(list(grp[\"ref_gene_id\"]))\n",
    "        assert len(rgids)==1,\"multiple group gids: \"+qgid+\"\\t\"+str(rgids)\n",
    "\n",
    "        known = len(pass_codes.intersection(set(list(grp[\"class_code\"]))))>0 # at least some codes in the passing codes - add everything to the known gene\n",
    "\n",
    "        rgid = \"\"\n",
    "        next_tid = 0\n",
    "        if not known: # novel\n",
    "            rgid = next_gid\n",
    "            next_gid+=1\n",
    "        else:\n",
    "            assert not rgid==\"-\",\"missing reference gid assigned\"\n",
    "            rgid = int(list(rgids)[0].split(\".\")[1])\n",
    "\n",
    "        for idx,row in grp.iterrows():\n",
    "            qtid = row[\"qry_id\"]\n",
    "            new_tid = \"\"\n",
    "            new_gid = \"\"\n",
    "            \n",
    "            if known:\n",
    "                code = row[\"class_code\"]\n",
    "                if code==\"=\": # already exists\n",
    "                    new_tid = row[\"ref_id\"]\n",
    "                    new_gid = row[\"ref_gene_id\"]\n",
    "                else: # add to known gene as a new transcript and increment max_tid\n",
    "                    # check if exists in the map\n",
    "                    if row[\"qry_id\"] in qtmap:\n",
    "                        new_tid = qtmap[row[\"qry_id\"]]\n",
    "                        new_gid = \"CHS.\"+new_tid.split(\".\")[1]\n",
    "                        gid_num = int(new_tid.split(\".\")[1])\n",
    "                        tid_num = int(new_tid.split(\".\")[2])\n",
    "#                         assert gid_num==rgid,\"gene ids do not match: \"+str(gid_num)+\" \"+str(rgid)+\" \"+\",\".join(rgids)+\" \"+new_tid\n",
    "                        assert new_tid not in loaded_tids,\"tid already found\"\n",
    "                    else:\n",
    "                        new_tid = \"CHS.\"+str(rgid)+\".\"+str(ref_gdict[rgid]+1)\n",
    "                        new_gid = \"CHS.\"+str(rgid)\n",
    "                        ref_gdict[rgid]+=1\n",
    "                \n",
    "            else:\n",
    "                # check if exists in the map\n",
    "                if row[\"qry_id\"] in qtmap:\n",
    "                    new_tid = qtmap[row[\"qry_id\"]]\n",
    "                    new_gid = \"CHS.\"+new_tid.split(\".\")[1]\n",
    "                else:\n",
    "                    new_tid = \"CHS.\"+str(rgid)+\".\"+str(next_tid)\n",
    "                    new_gid = \"CHS.\"+str(rgid)\n",
    "                    next_tid+=1\n",
    "            \n",
    "            if new_tid in loaded_tids:\n",
    "                print(\"duplicate ID: \"+new_tid)\n",
    "                continue\n",
    "            \n",
    "            loaded_tids.add(new_tid)\n",
    "            rec=qry_recs[qtid]\n",
    "            for line in rec.split(\"\\n\"):\n",
    "                if len(line.strip())==0:\n",
    "                    continue\n",
    "                lcs = line.strip().split(\"\\t\")\n",
    "                atts = [(x.split(\" \\\"\")[0].strip(),x.split(\"\\\"\")[1]) for x in lcs[8].strip(\";\").split(\";\")]\n",
    "                \n",
    "                for k,v in atts:\n",
    "                    if k==kv2add[0]:\n",
    "                        kv2add[1]+=\",\"+v\n",
    "                        \n",
    "                atts = [x for x in atts if x[0] not in [\"transcript_id\",\"gene_id\",kv2add[0]]]\n",
    "                        \n",
    "                atts = \"transcript_id \\\"\"+new_tid+\\\n",
    "                       \"\\\"; gene_id \\\"\"+new_gid+\\\n",
    "                       \"\\\"; old_transcript_id \\\"\"+row[\"qry_id\"]+\\\n",
    "                       \"\\\"; old_gene_id \\\"\"+row[\"qry_gene_id\"]+\\\n",
    "                       \"\\\"; \"+kv2add[0]+\" \\\"\"+kv2add[1]+\\\n",
    "                       \"\\\"; \"+\"; \".join([x[0]+\" \\\"\"+x[1]+\"\\\"\" for x in atts])+\";\"\n",
    "\n",
    "                lcs[8] = atts\n",
    "                outFP.write(\"\\t\".join(lcs)+\"\\n\")\n",
    "                \n",
    "                \n",
    "    # add the rest of the reference\n",
    "    with open(ref_gtf_fname,\"r\") as inFP:\n",
    "        for line in inFP:\n",
    "            if len(line.strip())==0:\n",
    "                continue\n",
    "            lcs = line.strip().split(\"\\t\")\n",
    "            tid = lcs[8].split(\"transcript_id \\\"\",1)[1].split(\"\\\"\",1)[0]\n",
    "            if not tid in loaded_tids:\n",
    "                outFP.write(line)\n",
    "                \n",
    "    outFP.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "a33a1787-a4ad-4a9c-8ed2-0e44c8b4cee8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load a map of \"=\" codes\n",
    "def load_trmap(tr_fname):\n",
    "    tqmap = dict()\n",
    "    qtmap = dict()\n",
    "    with open(tr_fname,\"r\") as inFP:\n",
    "        for line in inFP:\n",
    "            if line[0]==\"#\":\n",
    "                continue\n",
    "            lcs = line.strip(\"\\n\").split(\"\\t\")\n",
    "\n",
    "            code = lcs[3]\n",
    "            if code == \"=\":\n",
    "                rtid = lcs[2].split(\"|\")[1]\n",
    "                qtid = lcs[4].split(\"|\")[1]\n",
    "                tqmap.setdefault(rtid,list())\n",
    "                tqmap[rtid].append(qtid)\n",
    "                \n",
    "                qtmap.setdefault(qtid,list())\n",
    "                qtmap[qtid].append(rtid)\n",
    "                \n",
    "    for k,v in tqmap.items():\n",
    "        tqmap[k] = \",\".join(v)\n",
    "        \n",
    "    for k,v in qtmap.items():\n",
    "        qtmap[k] = \",\".join(v)\n",
    "        \n",
    "    return tqmap,qtmap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a9c9e279-3d17-4651-b448-8ed4e097dcdc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# run gffcompare between the must include and chess\n",
    "if not os.path.exists(out_dir+\"tmp_gen/\"):\n",
    "    os.makedirs(out_dir+\"tmp_gen/\")\n",
    "gffcmp_cmd = [gffcompare_bin,\"--no-merge\",\n",
    "              \"-o\", out_dir+\"tmp_gen/cmp\",\n",
    "              \"-r\", chess3_gtf_fname,\n",
    "              out_gencode_fname]\n",
    "print(\" \".join(gffcmp_cmd))\n",
    "subprocess.call(gffcmp_cmd)\n",
    "shutil.move(out_dir+\"cmp.must_include.gencode.gtf.refmap\",out_dir+\"tmp_gen/cmp.refmap\")\n",
    "shutil.move(out_dir+\"cmp.must_include.gencode.gtf.tmap\",out_dir+\"tmp_gen/cmp.tmap\")\n",
    "\n",
    "gffcmp_cmd = [gffcompare_bin,\"--no-merge\",\n",
    "              \"-o\", out_dir+\"tmp_gen/cmp_id\",\n",
    "              \"-r\", chess2_gtf_fname,\n",
    "              out_gencode_fname]\n",
    "print(\" \".join(gffcmp_cmd))\n",
    "subprocess.call(gffcmp_cmd)\n",
    "tqmap,qtmap = load_trmap(out_dir+\"tmp_gen/cmp_id.tracking\")\n",
    "\n",
    "add_from_tmap(chess3_gtf_fname,out_gencode_fname,out_dir+\"tmp_gen/cmp.tmap\",out_dir+\"tmp.ref_gen.gtf\",[\"step6\",\"GENCODE\"],qtmap)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f3596c73-bd6e-4ae2-9191-21b7094f2a33",
   "metadata": {},
   "outputs": [],
   "source": [
    "# repeat the same procedure with the new set of transcripts from refseq\n",
    "if not os.path.exists(out_dir+\"tmp_ref/\"):\n",
    "    os.makedirs(out_dir+\"tmp_ref/\")\n",
    "gffcmp_cmd = [gffcompare_bin,\"--no-merge\",\n",
    "              \"-o\", out_dir+\"tmp_ref/cmp\",\n",
    "              \"-r\", out_dir+\"tmp.ref_gen.gtf\",\n",
    "              out_refseq_fname]\n",
    "print(\" \".join(gffcmp_cmd))\n",
    "subprocess.call(gffcmp_cmd)\n",
    "shutil.move(out_dir+\"cmp.must_include.refseq.gtf.refmap\",out_dir+\"tmp_ref/cmp.refmap\")\n",
    "shutil.move(out_dir+\"cmp.must_include.refseq.gtf.tmap\",out_dir+\"tmp_ref/cmp.tmap\")\n",
    "\n",
    "gffcmp_cmd = [gffcompare_bin,\"--no-merge\",\n",
    "              \"-o\", out_dir+\"tmp_ref/cmp_id\",\n",
    "              \"-r\", chess2_gtf_fname,\n",
    "              out_refseq_fname]\n",
    "print(\" \".join(gffcmp_cmd))\n",
    "subprocess.call(gffcmp_cmd)\n",
    "tqmap,qtmap = load_trmap(out_dir+\"tmp_ref/cmp_id.tracking\")\n",
    "\n",
    "add_from_tmap(out_dir+\"tmp.ref_gen.gtf\",out_refseq_fname,out_dir+\"tmp_ref/cmp.tmap\",out_dir+\"tmp.ref_gen_refseq.gtf\",[\"step6\",\"RefSeq\"],qtmap)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ef9a9c1f-d69f-4136-9e7f-33d86e40ca43",
   "metadata": {},
   "outputs": [],
   "source": [
    "# repeat the same procedure with the new set of transcripts from mane\n",
    "if not os.path.exists(out_dir+\"tmp_mane/\"):\n",
    "    os.makedirs(out_dir+\"tmp_mane/\")\n",
    "gffcmp_cmd = [gffcompare_bin,\"--no-merge\",\n",
    "              \"-o\", out_dir+\"tmp_mane/cmp\",\n",
    "              \"-r\", out_dir+\"tmp.ref_gen_refseq.gtf\",\n",
    "              mane_gtf_fname]\n",
    "print(\" \".join(gffcmp_cmd))\n",
    "subprocess.call(gffcmp_cmd)\n",
    "mane_tmap = \"/\".join(mane_gtf_fname.split(\"/\")[:-1])+\"/cmp.\"+mane_gtf_fname.split(\"/\")[-1]+\".tmap\"\n",
    "mane_refmap = \"/\".join(mane_gtf_fname.split(\"/\")[:-1])+\"/cmp.\"+mane_gtf_fname.split(\"/\")[-1]+\".refmap\"\n",
    "shutil.move(mane_refmap,out_dir+\"tmp_mane/cmp.refmap\")\n",
    "shutil.move(mane_tmap,out_dir+\"tmp_mane/cmp.tmap\")\n",
    "\n",
    "gffcmp_cmd = [gffcompare_bin,\"--no-merge\",\n",
    "              \"-o\", out_dir+\"tmp_mane/cmp_id\",\n",
    "              \"-r\", chess2_gtf_fname,\n",
    "              mane_gtf_fname]\n",
    "print(\" \".join(gffcmp_cmd))\n",
    "subprocess.call(gffcmp_cmd)\n",
    "tqmap,qtmap = load_trmap(out_dir+\"tmp_mane/cmp_id.tracking\")\n",
    "\n",
    "add_from_tmap(out_dir+\"tmp.ref_gen_refseq.gtf\",mane_gtf_fname,out_dir+\"tmp_mane/cmp.tmap\",out_dir+\"tmp.ref_gen_refseq_mane.gtf\",[\"step6\",\"MANE\"],qtmap)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "83330fa0-967d-4f0f-98a1-51a9f0143f25",
   "metadata": {},
   "outputs": [],
   "source": [
    "# lastly, we need to add ALL-ids for all transcripts based on the comparison to the 27M set\n",
    "\n",
    "gffcmp_cmd = [\"gffcompare\",\n",
    "              \"-o\",out_dir+\"27McmpStep6\",\n",
    "              \"-r\",out_dir+\"tmp.ref_gen_refseq_mane.gtf\",\n",
    "              M27_gtf_fname]\n",
    "print(\" \".join(gffcmp_cmd))\n",
    "# subprocess.call(gffcmp_cmd)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "6713ae5d-c3eb-4074-b42f-c1c7e112b90c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load a map of IDs\n",
    "ass_id_map = dict()\n",
    "with open(out_dir+\"27McmpStep6.tracking\",\"r\") as inFP:\n",
    "    for line in inFP:\n",
    "        lcs = line.split(\"\\t\")\n",
    "        code = lcs[3]\n",
    "        if code == \"=\":\n",
    "            ass_id = lcs[4].split(\"|\")[1]\n",
    "            chs_id = lcs[2].split(\"|\")[1]\n",
    "            ass_id_map[chs_id] = ass_id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "fbb2a288-e692-4d22-9097-338204bb2acb",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(out_dir+\"chess.gtf\",\"w+\") as outFP:\n",
    "    with open(out_dir+\"tmp.ref_gen_refseq_mane.gtf\",\"r\") as inFP:\n",
    "        for line in inFP:\n",
    "            if len(line.strip())==0:\n",
    "                continue\n",
    "            lcs = line.strip().split(\"\\t\")\n",
    "            if len(lcs)<9:\n",
    "                continue\n",
    "            if lcs[2]==\"transcript\":\n",
    "                # add attribute with the original ID\n",
    "                tid = lcs[8].split(\"transcript_id \\\"\",1)[1].split(\"\\\"\",1)[0]\n",
    "                ass_id = \"none\"\n",
    "                if tid in ass_id_map:\n",
    "                    ass_id = ass_id_map[tid]\n",
    "                lcs[8] = lcs[8].rstrip().rstrip(\";\")+\"; assembly_id \\\"\"+ass_id+\"\\\";\"\n",
    "            line = \"\\t\".join(lcs).rstrip()+\"\\n\"\n",
    "            outFP.write(line)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "770173ee-a1fe-4008-bcf2-b3b9b64f9844",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
