{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "4402ec5f-c1d3-49ea-b5e1-b17355e94d24",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import subprocess\n",
    "import random\n",
    "from scipy import stats\n",
    "import pickle\n",
    "import glob\n",
    "import math\n",
    "import csv\n",
    "import sys\n",
    "import os\n",
    "import re\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib import animation\n",
    "import seaborn as sns\n",
    "plt.rcParams['figure.figsize'] = (20.0, 10.0)\n",
    "plt.rcParams['font.family'] = \"serif\"\n",
    "plt.rcParams['font.size'] = 24\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "4eae9456-e858-4750-b4c0-f5b4e09cf2af",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The autoreload extension is already loaded. To reload it, use:\n",
      "  %reload_ext autoreload\n"
     ]
    }
   ],
   "source": [
    "pd.set_option('display.max_columns', None)\n",
    "\n",
    "%load_ext autoreload\n",
    "%autoreload 1\n",
    "\n",
    "sys.path.insert(0, \"./soft\")\n",
    "%aimport definitions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "9a2213e0-45d1-4c22-9ec7-94d4ac7e2daf",
   "metadata": {},
   "outputs": [],
   "source": [
    "base_dir = \"./\"\n",
    "data_dir = base_dir+\"data/\"\n",
    "chess3_data_dir = data_dir+\"chess3/\"\n",
    "\n",
    "gff3cols = [\"seqid\",\"source\",\"type\",\"start\",\"end\",\"score\",\"strand\",\"phase\",\"attributes\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "a0620fe8-a162-46f4-9492-e47537037cec",
   "metadata": {},
   "outputs": [],
   "source": [
    "chess30_gtf_fname = chess3_data_dir+\"chess3.0.gtf\"\n",
    "chess31_gtf_fname = chess3_data_dir+\"chess3.1.gtf\"\n",
    "chess31_gff_fname = chess3_data_dir+\"chess3.1.gff\"\n",
    "\n",
    "chess30_chm13_gtf_fname = chess3_data_dir+\"chess3.0.CHM13.gtf\"\n",
    "chess31_chm13_gtf_fname = chess3_data_dir+\"chess3.1.CHM13.gtf\"\n",
    "chess31_chm13_gff_fname = chess3_data_dir+\"chess3.1.CHM13.gff\"\n",
    "\n",
    "chess2_gtf_fname = data_dir+\"chess2.2.gtf\"\n",
    "all_gtf_fname = data_dir+\"all.gtf\"\n",
    "tb_gtf_fname = data_dir+\"assembled.gtf\"\n",
    "refseq_gtf_fname = data_dir+\"refseq.gtf\" # 110 - same as preprint\n",
    "refseq_gff_fname = data_dir+\"refseq.gff\"\n",
    "gencode_gtf_fname = data_dir+\"gencode.gtf\" # 41 - same as preprint\n",
    "\n",
    "chess2_chess3_mappings_fname = \"./chess2_and_chess3.mappings\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "188890da-9259-4269-9647-b96d75154887",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get latest release from https://github.com/chess-genome/chess/archive/refs/tags/v.3.0.zip"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "4c3cc032-082e-468c-a179-dcf7a4352763",
   "metadata": {},
   "outputs": [],
   "source": [
    "def extract_attributes(attribute_str:str,gff=False)->dict: # extract attribute key values into dictionary\n",
    "    attrs = attribute_str.rstrip().rstrip(\";\").split(\";\")\n",
    "    attrs = [x.strip() for x in attrs]\n",
    "    attrs = [x.strip(\"\\\"\") for x in attrs]\n",
    "    attrs_dict = dict()\n",
    "    sep = \" \\\"\"\n",
    "    if gff:\n",
    "        sep = \"=\"\n",
    "    for at in attrs:\n",
    "        k,v = at.split(sep)\n",
    "        attrs_dict.setdefault(k,v)\n",
    "        \n",
    "    return attrs_dict\n",
    "\n",
    "def rename_attributes(attrs:dict)->dict: # renames attributes\n",
    "    rename_dict = {\"max_tpm\":\"max_TPM\",\n",
    "                   \"num_samples\":\"sample_count\"}\n",
    "    res_dict = {}\n",
    "    for k,v in attrs.items():\n",
    "        if k in rename_dict:\n",
    "            res_dict[rename_dict[k]]=v\n",
    "        else:\n",
    "            res_dict[k]=v\n",
    "    return res_dict\n",
    "\n",
    "def convert_source(source: str,db_xref: bool, is_mane: bool)->str: # take line in and replace source according to hard-coded dictionary\n",
    "    kvs = {\"BestRefSeq\":     \"RefSeq\",\n",
    "           \"Curated Genomic\":\"RefSeq\",\n",
    "           \"RefSeq\":         \"RefSeq\",\n",
    "           \"Gnomon\":         \"RefSeq\",\n",
    "           \"tRNAscan-SE\":    \"RefSeq\",\n",
    "           \"cmsearch\":       \"RefSeq\",\n",
    "           \"ENSEMBL\":        \"GENCODE\",\n",
    "           \"HAVANA\":         \"GENCODE\",\n",
    "           \"ensembl\":        \"GENCODE\",\n",
    "           \"ensembl_havana\": \"GENCODE\",\n",
    "           \"havana\":         \"GENCODE\",\n",
    "           \"FANTOM\":         \"CHESS\",\n",
    "           \"StringTie\":      \"CHESS\",\n",
    "           \"Liftoff\":        \"Liftoff\"}\n",
    "    \n",
    "    if is_mane:\n",
    "        return \"MANE\"\n",
    "    \n",
    "    if not db_xref: # anything without \"db_xref\" needs to be assigned CHESS type\n",
    "        return \"CHESS\"\n",
    "    \n",
    "    assert source in kvs,\"unknown source: \"+source\n",
    "    return kvs[source]\n",
    "\n",
    "def get_source_conversion(in_gtf_fname: str) -> dict():\n",
    "    res = dict()\n",
    "    with open(in_gtf_fname,\"r\") as inFP:\n",
    "        for line in inFP:\n",
    "            if line[0]==\"#\": # comment found\n",
    "                outFP.write(line)\n",
    "\n",
    "            lcs = line.split(\"\\t\")\n",
    "            if not len(lcs) == 9:\n",
    "                continue\n",
    "    \n",
    "            if not lcs[2]==\"transcript\":\n",
    "                continue\n",
    "\n",
    "            attrs = extract_attributes(lcs[8])\n",
    "            attrs = rename_attributes(attrs)\n",
    "\n",
    "            # convert source\n",
    "            is_mane = False\n",
    "            if \"tag\" in attrs:\n",
    "                is_mane = \"mane\" in attrs[\"tag\"].lower()\n",
    "            new_source = convert_source(lcs[1],\"db_xref\" in attrs,is_mane)\n",
    "            res[attrs[\"transcript_id\"]] = new_source\n",
    "            \n",
    "    return res\n",
    "\n",
    "def to_attribute_string(attrs:dict,gff=False,feature_type=None)->str: # converts attribute key values back into string\n",
    "    order = [\"ID\",\"Parent\",\"transcript_id\",\"gene_id\",\"gene_name\",\"gene_type\",\"db_xref\",\"description\",\"max_TPM\",\"sample_count\",\"assembly_id\",\"tag\"]\n",
    "    res = \"\"\n",
    "    sep = \" \"\n",
    "    quote = \"\\\"\"\n",
    "    end = \"; \"\n",
    "    if gff:\n",
    "        assert feature_type in [\"gene\",\"transcript\",\"exon\",\"CDS\"],\"wrong type: \"+str(feature_type)\n",
    "        sep = \"=\"\n",
    "        quote = \"\"\n",
    "        end = \";\"\n",
    "        \n",
    "    for k in order:\n",
    "        if k in attrs:\n",
    "            if gff:\n",
    "                assert \";\" not in attrs[k],\"invalid character in attribute: \"+attrs[k]\n",
    "            \n",
    "            if gff and \n",
    "            \n",
    "            if gff and feature_type==\"gene\" and k==\"transcript_id\":\n",
    "                continue\n",
    "            elif gff and feature_type==\"gene\" and k==\"gene_id\":\n",
    "                res+=\"ID=\"+quote+attrs[k]+quote+end\n",
    "            elif gff and feature_type==\"transcript\" and k==\"transcript_id\":\n",
    "                res+=\"ID=\"+quote+attrs[k]+quote+end\n",
    "            elif gff and feature_type==\"transcript\" and k==\"gene_id\":\n",
    "                res+=\"Parent=\"+quote+attrs[k]+quote+end\n",
    "            elif gff and feature_type in [\"exon\",\"CDS\"] and k==\"transcript_id\":\n",
    "                res+=\"Parent=\"+quote+attrs[k]+quote+end\n",
    "            elif gff and feature_type in [\"exon\",\"CDS\"] and k==\"gene_id\":\n",
    "                continue\n",
    "            else:        \n",
    "                res+=k+sep+quote+attrs[k]+quote+end\n",
    "    \n",
    "    # add any other attributes in sorted order\n",
    "    for k in sorted(list(attrs)):\n",
    "        if k not in order:\n",
    "            if gff:\n",
    "                assert \";\" not in attrs[k],\"invalid character in attribute: \"+attrs[k]\n",
    "            res+=k.lower()+sep+quote+attrs[k]+quote+end\n",
    "    \n",
    "    if not gff:\n",
    "        res = res.rstrip()\n",
    "    if gff:\n",
    "        res = res.rstrip(\";\")\n",
    "    return res\n",
    "    \n",
    "def round_tpm(tpm_str:str)->str:\n",
    "    tpm = 0\n",
    "    try:\n",
    "        tpm = float(tpm_str)\n",
    "    except:\n",
    "        print(\"unable to convert TPM: \"+tpm_str)\n",
    "    \n",
    "    if tpm<0.01:\n",
    "        return tpm_str\n",
    "    \n",
    "    if tpm<0.1:\n",
    "        return str(round(tpm,2))\n",
    "        \n",
    "    if tpm<5:\n",
    "        return str(round(tpm,1))\n",
    "    \n",
    "    else:\n",
    "        return str(round(tpm))\n",
    "    \n",
    "    \n",
    "def get_comments()->str:\n",
    "    comment_str = \"##NAME: CHESS\\n\"+\\\n",
    "                  \"##VERSION: 3.01\\n\"+\\\n",
    "                  \"##DESCRIPTION: Comprehensive Human Expressed SequenceS\\n\"+\\\n",
    "                  \"##FORMAT: GTF\\n\"+\\\n",
    "                  \"##CONTACT: ales[dot]varabyou[at]jhu[dot][edu],mpertea[at]jhu[dot][edu]\\n\"\n",
    "    return comment_str\n",
    "\n",
    "def get_assembly_id(attrs:dict)->str:\n",
    "    if \"assembly_id\" in attrs:\n",
    "        return attrs[\"assembly_id\"]\n",
    "    else:\n",
    "        return \"NA\"\n",
    "        \n",
    "\n",
    "def gtf_fixes(in_gtf_fname:str,out_gtf_fname:str): # apply fixes to the input GTF and write to output\n",
    "    \n",
    "    # collect a dictionary of tids to sources\n",
    "    tid2source = get_source_conversion(in_gtf_fname)\n",
    "    \n",
    "    assert os.path.exists(in_gtf_fname),\"Input GTF file not found: \"+in_gtf_fname\n",
    "    \n",
    "    # get transcript ids to exclude\n",
    "    tids_to_remove = definitions.get_attribute(in_gtf_fname,\"gene_type\")\n",
    "    tids_to_remove = set(tids_to_remove[tids_to_remove[\"gene_type\"]==\"other\"][\"tid\"])\n",
    "    print(\"number of transcripts being removed: \"+str(len(tids_to_remove)))\n",
    "\n",
    "    with open(out_gtf_fname,\"w+\") as outFP:\n",
    "        outFP.write(get_comments())\n",
    "        with open(in_gtf_fname,\"r\") as inFP:\n",
    "            for line in inFP:\n",
    "                if line[0]==\"#\": # comment found\n",
    "                    outFP.write(line)\n",
    "                    \n",
    "                lcs = line.split(\"\\t\")\n",
    "                if not len(lcs) == 9:\n",
    "                    continue\n",
    "                    \n",
    "                attrs = extract_attributes(lcs[8])\n",
    "                attrs = rename_attributes(attrs)\n",
    "                if attrs[\"transcript_id\"] in tids_to_remove:\n",
    "                    continue\n",
    "                \n",
    "                # setup copies for output\n",
    "                res_lcs = lcs[:8]\n",
    "                res_attrs = attrs\n",
    "                \n",
    "                # convert source\n",
    "                res_lcs[1] = tid2source[attrs[\"transcript_id\"]]\n",
    "                \n",
    "                # add source information to the attributes\n",
    "                if lcs[2]==\"transcript\":\n",
    "                    res_attrs[\"original_source\"]=lcs[1]\n",
    "                \n",
    "                    # round TPM\n",
    "                    if \"max_TPM\" in attrs:\n",
    "                        res_attrs[\"max_TPM\"] = round_tpm(res_attrs[\"max_TPM\"])\n",
    "                        \n",
    "                    res_attrs[\"assembly_id\"] = get_assembly_id(res_attrs)\n",
    "                \n",
    "                if lcs[2] in [\"CDS\",\"exon\"]: # remove extra attribtues\n",
    "                    for k in list(res_attrs):\n",
    "                        if k not in [\"transcript_id\",\"gene_id\",\"gene_name\"]:\n",
    "                            del res_attrs[k]\n",
    "                    \n",
    "                    \n",
    "                res_line = \"\\t\".join(res_lcs)+\"\\t\"+to_attribute_string(res_attrs)\n",
    "                outFP.write(res_line+\"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "f883373a-cff4-4262-b106-b84e8f7b3c29",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of transcripts being removed: 24\n"
     ]
    }
   ],
   "source": [
    "gtf_fixes(chess30_gtf_fname,chess31_gtf_fname)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "95d3dc5e-772c-45a4-8564-b168052b34cf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tid</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_type</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>other</th>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pseudogene</th>\n",
       "      <td>16521</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              tid\n",
       "gene_type        \n",
       "other          16\n",
       "pseudogene  16521"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# we need to remove converted gene records from refseq\n",
    "gene_names = definitions.get_attribute(chess30_gtf_fname,[\"gene_name\",\"db_xref\",\"assembly_id\",\"gene_type\",\"gene_id\"])\n",
    "gene_names[\"refseq\"] = gene_names[\"db_xref\"].str.split(\"RefSeq:\",expand=True)[1].str.split(\",\",expand=True)[0]\n",
    "\n",
    "tmp = definitions.get_chains(chess30_gtf_fname,\"exon\",True)\n",
    "tmp[\"exon_count\"] = tmp[\"chain\"].apply(lambda x: len(x))\n",
    "gene_names = gene_names.merge(tmp[[\"tid\",\"exon_count\"]],on=\"tid\",how=\"left\")\n",
    "gene_names = gene_names[(gene_names[\"gene_name\"]==gene_names[\"refseq\"])&\\\n",
    "                        (gene_names[\"exon_count\"]==1)&\\\n",
    "                        (gene_names[\"assembly_id\"].isin([\"-\",\"NA\"]))].reset_index(drop=True)\n",
    "gene_names[[\"gene_type\",\"tid\"]].groupby(by=\"gene_type\").count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "db5df436-0b45-4273-8c48-232a06de86b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "pseudogene_set = set(gene_names[gene_names[\"gene_type\"]==\"pseudogene\"][\"gene_id\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "caca4836-3a04-4032-bb65-6b2c729b23b5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MIR6859-1</td>\n",
       "      <td>microRNA 6859-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MIR1302-2</td>\n",
       "      <td>microRNA 1302-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>FAM138A</td>\n",
       "      <td>family with sequence similarity 138 member A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>OR4F5</td>\n",
       "      <td>olfactory receptor family 4 subfamily F member 5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>LOC729737</td>\n",
       "      <td>uncharacterized LOC729737</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        Name                                       description\n",
       "0  MIR6859-1                                   microRNA 6859-1\n",
       "1  MIR1302-2                                   microRNA 1302-2\n",
       "2    FAM138A      family with sequence similarity 138 member A\n",
       "3      OR4F5  olfactory receptor family 4 subfamily F member 5\n",
       "4  LOC729737                         uncharacterized LOC729737"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get  gene names from the refseq gtf\n",
    "refseq_gene_desc = definitions.get_attribute(refseq_gff_fname,[\"Name\",\"description\"],cols=None,feature=\"gene\",gff=True)\n",
    "refseq_gene_desc = refseq_gene_desc[refseq_gene_desc[\"description\"]!=\"-\"][[\"Name\",\"description\"]].reset_index(drop=True)\n",
    "refseq_gene_desc.drop_duplicates(inplace=True)\n",
    "refseq_gene_desc_dict = refseq_gene_desc.set_index('Name').to_dict()['description']\n",
    "refseq_gene_desc.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "a8d77e90-fcee-4d75-8e6c-da3337051bad",
   "metadata": {},
   "outputs": [],
   "source": [
    "# add gene records to gtf\n",
    "\n",
    "# collect all transcripts by their gene_id\n",
    "\n",
    "genes = dict()\n",
    "\n",
    "with open(chess31_gtf_fname,\"r\") as inFP:\n",
    "    for line in inFP:\n",
    "        if line[0]==\"#\": # comment found\n",
    "            continue\n",
    "\n",
    "        lcs = line.split(\"\\t\")\n",
    "        if not len(lcs) == 9:\n",
    "            continue\n",
    "\n",
    "        attrs = extract_attributes(lcs[8])\n",
    "        gid = attrs[\"gene_id\"]\n",
    "        tid = attrs[\"transcript_id\"]\n",
    "        \n",
    "        if lcs[2]==\"transcript\":\n",
    "            genes.setdefault(gid,[dict({\"seqid\":None,\n",
    "                                        \"start\":sys.maxsize,\n",
    "                                         \"end\":0,\n",
    "                                         \"strand\":None,\n",
    "                                         \"name\":set(),\n",
    "                                         \"type\":set()}),dict()])\n",
    "            genes[gid][0][\"start\"] = min(genes[gid][0][\"start\"],int(lcs[3]))\n",
    "            genes[gid][0][\"end\"] = max(genes[gid][0][\"end\"],int(lcs[4]))\n",
    "            \n",
    "            if not genes[gid][0][\"seqid\"] is None:\n",
    "                assert genes[gid][0][\"seqid\"] == lcs[0],\"wrong seqid: \"+gid\n",
    "            else:\n",
    "                genes[gid][0][\"seqid\"] = lcs[0]\n",
    "\n",
    "            assert lcs[6] in [\"+\",\"-\"],\"wrong strand: \"+tid+\" : \"+lcs[6]\n",
    "            if not genes[gid][0][\"strand\"] is None:\n",
    "                if genes[gid][0][\"strand\"] != lcs[6]:\n",
    "                    genes[gid][0][\"strand\"] = \".\"\n",
    "            else:\n",
    "                genes[gid][0][\"strand\"] = lcs[6]\n",
    "\n",
    "            if \"gene_name\" in attrs:\n",
    "                genes[gid][0][\"name\"].add(attrs[\"gene_name\"])\n",
    "\n",
    "            if \"gene_type\" in attrs:\n",
    "                genes[gid][0][\"type\"].add(attrs[\"gene_type\"])\n",
    "\n",
    "            genes[gid][1].setdefault(tid,\"\")\n",
    "            \n",
    "        assert tid in genes[gid][1],\"wrong tid (unsorted?): \"+tid\n",
    "        genes[gid][1][tid]+=line\n",
    "        \n",
    "        \n",
    "# get minimum and maximum coordinates\n",
    "       \n",
    "with open(chess31_gtf_fname+\".with_genes.gtf\",\"w+\") as outFP:\n",
    "    # write headers\n",
    "    with open(chess31_gtf_fname,\"r\") as inFP:\n",
    "        for line in inFP:\n",
    "            if line[0]==\"#\": # comment found\n",
    "                outFP.write(line)\n",
    "            else: # break at the first non-header line\n",
    "                break\n",
    "            \n",
    "    for gid,gv in genes.items():\n",
    "        gline = gv[0][\"seqid\"]+\"\\t\"+\\\n",
    "                    \"CHESS\"+\"\\t\"+\\\n",
    "                    \"gene\"+\"\\t\"+\\\n",
    "                    str(gv[0][\"start\"])+\"\\t\"+\\\n",
    "                    str(gv[0][\"end\"])+\"\\t\"+\\\n",
    "                    \".\"+\"\\t\"+\\\n",
    "                    gv[0][\"strand\"]+\"\\t\"+\\\n",
    "                    \".\"+\"\\t\"+\\\n",
    "                    \"gene_id \\\"\"+gid+\"\\\";\"\n",
    "        \n",
    "        \n",
    "        if len(gv[0][\"type\"])>0:\n",
    "            gline+=\" gene_type \\\"\"+\", \".join(list(gv[0][\"type\"]))+\"\\\";\"\n",
    "        \n",
    "        if len(gv[0][\"name\"])>0:\n",
    "            gline+=\" gene_name \\\"\"+\", \".join(list(gv[0][\"name\"]))+\"\\\";\"\n",
    "            desc = list()\n",
    "            for n in gv[0][\"name\"]:\n",
    "                if n in refseq_gene_desc_dict:\n",
    "                    desc.append([n,refseq_gene_desc_dict[n]])\n",
    "                    \n",
    "            if len(desc)==1:\n",
    "                gline+=\" description \\\"\"+desc[0][1]+\"\\\";\"\n",
    "            if len(desc)>1:\n",
    "                print(gv[0][\"name\"])\n",
    "                gline+=\" description \\\"\"+\", \".join([x[0]+\" : \"+x[1] for x in desc])+\"\\\";\"\n",
    "                \n",
    "        outFP.write(gline+\"\\n\")        \n",
    "        if not gid in pseudogene_set: # only write transcript features for non-pseudogene entries\n",
    "            for tid,tv in gv[1].items():\n",
    "                outFP.write(tv)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "400dbd67-b2f8-4af7-829c-dd6c29cfa6c4",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "e22cc29d-d2a0-4a26-b20d-33ff6f809a50",
   "metadata": {},
   "outputs": [],
   "source": [
    "# work on attribtue stats\n",
    "\n",
    "def extract_all_attributes(gtf_fname: str, feature_type: str, gff3: bool=False) -> dict:\n",
    "    sep = \" \\\"\"\n",
    "    if gff3:\n",
    "        sep = \"=\"\n",
    "    \n",
    "    res = dict()\n",
    "    \n",
    "    assert os.path.exists(gtf_fname),\"the GTF file not found: \"+gtf_fname\n",
    "    with open(gtf_fname,\"r\") as inFP:\n",
    "        for line in inFP:\n",
    "            lcs = line.split(\"\\t\")\n",
    "            if not len(lcs) == 9:\n",
    "                continue\n",
    "                \n",
    "            if lcs[2] != feature_type:\n",
    "                continue\n",
    "                \n",
    "            attrs = lcs[8].strip().strip(\";\").split(\";\")\n",
    "            attrs = [x.strip() for x in attrs]\n",
    "            attrs = [x.strip(\"\\\"\") for x in attrs]\n",
    "            for at in attrs:\n",
    "                k,v = at.split(sep)\n",
    "                res.setdefault(k,dict())\n",
    "                res[k].setdefault(v,0)\n",
    "                res[k][v]+=1\n",
    "                \n",
    "    return res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "2a64175a-ab1b-44a2-b926-9b126e766857",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['transcript_id', 'gene_id', 'gene_name', 'gene_type', 'db_xref', 'assembly_id', 'original_source', 'max_TPM', 'sample_count', 'tag']\n",
      "{'transcribed_pseudogene': 1954, 'miRNA': 5151, 'lncRNA': 36356, 'pseudogene': 16521, 'protein_coding': 105328, 'snRNA': 159, 'snoRNA': 1251, 'antisense_RNA': 37, 'ncRNA': 28, 'tRNA': 643, 'TEC': 28, 'ncRNA_pseudogene': 1, 'misc_RNA': 83, 'V_segment': 342, 'rRNA': 40, 'C_region': 36, 'J_segment': 117, 'V_segment_pseudogene': 282, 'telomerase_RNA': 1, 'vault_RNA': 4, 'D_segment': 61, 'J_segment_pseudogene': 11, 'Y_RNA': 4, 'RNase_MRP_RNA': 1, 'scRNA': 4, 'RNase_P_RNA': 1, 'C_region_pseudogene': 7}\n",
      "{'MANE_Select': 19043, 'partial': 1368, 'MANE_Select,duplicated_transcript': 77, 'duplicated_transcript': 175}\n",
      "{'BestRefSeq': 72352, 'HAVANA': 8047, 'Curated Genomic': 16319, 'havana': 4137, 'Gnomon': 16555, 'ensembl_havana': 14981, 'cmsearch': 1184, 'StringTie': 33783, 'FANTOM': 318, 'tRNAscan-SE': 621, 'ensembl': 2, 'ENSEMBL': 115, 'RefSeq': 37}\n"
     ]
    }
   ],
   "source": [
    "c3a_gtf = extract_all_attributes(chess31_gtf_fname,\"transcript\")\n",
    "print(list(c3a_gtf))\n",
    "print(c3a_gtf[\"gene_type\"])\n",
    "print(c3a_gtf[\"tag\"])\n",
    "print(c3a_gtf[\"original_source\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d5465023-639e-41b2-91b9-a8c0dfdd46f3",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "c6781cfb-5379-4252-ab35-021be15576fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "# need to convert to GFF and other formats now\n",
    "def gtf2gff(gtf_fname:str,gff_fname:str):\n",
    "    with open(gff_fname,\"w+\") as outFP:\n",
    "        outFP.write(\"##gff-version 3\\n\")\n",
    "        outFP.write(\"#!gff-spec-version 1.21\\n\")\n",
    "        with open(gtf_fname,\"r\") as inFP:\n",
    "            for line in inFP:\n",
    "                lcs = line.split(\"\\t\")\n",
    "                if lcs[0]==\"#\":\n",
    "                    outFP.write(line)\n",
    "                if not len(lcs) == 9:\n",
    "                    continue\n",
    "                    \n",
    "                attrs = extract_attributes(lcs[8])\n",
    "                res_line = \"\\t\".join(lcs[:-1])+\"\\t\"+to_attribute_string(attrs,True,lcs[2])\n",
    "                outFP.write(res_line+\"\\n\")\n",
    "                \n",
    "gtf2gff(chess31_gtf_fname+\".with_genes.gtf\",chess31_gff_fname+\".with_genes.gff\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "8be98787-101a-4716-8c3a-c84238285910",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of transcripts being removed: 14\n"
     ]
    }
   ],
   "source": [
    "# apply fixes and other corrections to the CHM13 annotation\n",
    "gtf_fixes(chess30_chm13_gtf_fname,chess31_chm13_gtf_fname)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bcf33688-279f-41d3-bbd7-b8e54ff54a0e",
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf2gff(chess31_chm13_gtf_fname,chess31_chm13_gff_fname)\n",
    "cmd = [\"gffread\",\"--keep-genes\",\"--keep-comments\",\"-F\",\"-O\",\"-o\",chess31_chm13_gff_fname.rstrip(\".gff\")+\".tmp.gff\",chess31_chm13_gff_fname]\n",
    "print(\" \".join(cmd))\n",
    "subprocess.call(cmd)\n",
    "# rename unwanted tags\n",
    "with open(chess31_chm13_gff_fname,\"w+\") as outFP:\n",
    "    with open(chess31_chm13_gff_fname.rstrip(\".gff\")+\".tmp.gff\",\"r\") as inFP:\n",
    "        for line in inFP:\n",
    "            lcs = line.split(\"\\t\")\n",
    "            if not len(lcs) == 9:\n",
    "                outFP.write(line)\n",
    "                continue\n",
    "                \n",
    "            attrs = lcs[8].rstrip().rstrip(\";\").split(\";\")\n",
    "            attrs = [x.strip() for x in attrs]\n",
    "            attrs = [x.strip(\"\\\"\") for x in attrs]\n",
    "            attrs = [x.split(\"=\") for x in attrs]\n",
    "            attr_str = \"\"\n",
    "            for k,v in attrs:\n",
    "                if k.lower() in [\"name\"]:\n",
    "                    continue\n",
    "                \n",
    "                attr_str+=k+\"=\"+v+\";\"\n",
    "            attr_str.rstrip(\";\")\n",
    "                \n",
    "            res_line = \"\\t\".join(lcs[:-1])+\"\\t\"+attr_str\n",
    "            outFP.write(res_line+\"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "587d6492-27bc-4d3d-91fd-28f6e54afb98",
   "metadata": {},
   "outputs": [],
   "source": [
    "## convert to GTF including the gene features\n",
    "with open(chess31_chm13_gtf_fname,\"w+\") as outFP:\n",
    "    with open(chess31_chm13_gff_fname,\"r\") as inFP:\n",
    "        for line in inFP:\n",
    "            lcs = line.split(\"\\t\")\n",
    "            if lcs[0]==\"#\":\n",
    "                if \"gff-version\" in lcs[0]:\n",
    "                    continue\n",
    "                if \"gff-spec-version\" in lcs[0]:\n",
    "                    continue\n",
    "                outFP.write(line)\n",
    "            if not len(lcs) == 9:\n",
    "                continue\n",
    "\n",
    "            attrs = extract_attributes(lcs[8],True)\n",
    "            res_attrs = dict()\n",
    "            for k,v in attrs.items():\n",
    "                if k.lower() in [\"parent\",\"id\"]:\n",
    "                    if k.lower()==\"parent\" and lcs[2]==\"gene\":\n",
    "                        print(\"error: \"+attrs)\n",
    "                        continue\n",
    "                    if k.lower()==\"parent\" and lcs[2]==\"transcript\":\n",
    "                        res_attrs[\"gene_id\"] = v\n",
    "                    if k.lower()==\"parent\" and lcs[2] in [\"CDS\",\"exon\"]:\n",
    "                        res_attrs[\"transcript_id\"] = v\n",
    "\n",
    "                    if k.lower()==\"id\" and lcs[2]==\"gene\":\n",
    "                        res_attrs[\"gene_id\"] = v\n",
    "                    if k.lower()==\"id\" and lcs[2]==\"transcript\":\n",
    "                        res_attrs[\"transcript_id\"] = v\n",
    "                    if k.lower()==\"id\" and lcs[2] in [\"CDS\",\"exon\"]:\n",
    "                        continue\n",
    "                else:\n",
    "                    res_attrs[k]=v\n",
    "                \n",
    "            res_line = \"\\t\".join(lcs[:-1])+\"\\t\"+to_attribute_string(res_attrs,False,lcs[2])\n",
    "            outFP.write(res_line+\"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "af501922-d8e5-4fd7-b4ab-402ed9cd8a56",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "619f67b6-23f5-4e5d-9551-be7b0564c671",
   "metadata": {},
   "outputs": [],
   "source": [
    "os.environ['LD_LIBRARY_PATH'] = \"/ccb/sw/lib/:LD_LIBRARY_PATH\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "a912a579-8533-4662-b5ac-833b0226608e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bedToBigBed -type=bed12+8 -tab -as=bigGenePred.as /ccb/salz8-1/avaraby/chess3_rerun_31102021/reviews/data/chess3/chess3.1.bedPlus /ccb/salz8-1/avaraby/chess3_rerun_31102021/reviews/data/chess3/hs38DH.len /ccb/salz8-1/avaraby/chess3_rerun_31102021/reviews/data/chess3/chess3.1.bb -extraIndex=name\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "pass1 - making usageList (307 chroms): 86 millis\n",
      "pass2 - checking and writing primary data (151915 records, 20 fields): 1326 millis\n",
      "Sorting and writing extra index 0: 34 millis\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# generate bedplus files\n",
    "cmd = [\"gff3ToGenePred\",chess31_gff_fname+\".with_genes.gff\",chess31_gff_fname.rstrip(\".gff\")+\".genePred\"]\n",
    "print(\" \".join(cmd))\n",
    "subprocess.call(cmd)\n",
    "\n",
    "cmd = [\"genePredToBigGenePred\",chess31_gff_fname.rstrip(\".gff\")+\".genePred\",chess31_gff_fname.rstrip(\".gff\")+\".bedPlus\"]\n",
    "print(\" \".join(cmd))\n",
    "subprocess.call(cmd)\n",
    "\n",
    "cmd = [\"bedSort\",chess31_gff_fname.rstrip(\".gff\")+\".bedPlus\",chess31_gff_fname.rstrip(\".gff\")+\".srt.bedPlus\"]\n",
    "print(\" \".join(cmd))\n",
    "subprocess.call(cmd)\n",
    "\n",
    "os.rename(chess31_gff_fname.rstrip(\".gff\")+\".srt.bedPlus\",chess31_gff_fname.rstrip(\".gff\")+\".bedPlus\")\n",
    "\n",
    "# cmd = [\"wget\",\"http://genome.ucsc.edu/goldenPath/help/examples/bigGenePred.as\"]\n",
    "# print(\" \".join(cmd))\n",
    "# subprocess.call(cmd)\n",
    "\n",
    "cmd = [\"bedToBigBed\",\"-type=bed12+8\",\"-tab\",\"-as=bigGenePred.as\",chess31_gff_fname.rstrip(\".gff\")+\".bedPlus\",chess3_data_dir+\"hs38DH.len\",chess31_gff_fname.rstrip(\".gff\")+\".bb\",\"-extraIndex=name\"]\n",
    "print(\" \".join(cmd))\n",
    "subprocess.call(cmd)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a4579a65-ab2a-4faf-9f17-b3df28c1f0b4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# same for CHM13\n",
    "# generate bedplus files\n",
    "cmd = [\"gff3ToGenePred\",chess31_chm13_gff_fname,chess31_chm13_gff_fname.rstrip(\".gff\")+\".genePred\"]\n",
    "print(\" \".join(cmd))\n",
    "subprocess.call(cmd)\n",
    "\n",
    "cmd = [\"genePredToBigGenePred\",chess31_chm13_gff_fname.rstrip(\".gff\")+\".genePred\",chess31_chm13_gff_fname.rstrip(\".gff\")+\".bedPlus\"]\n",
    "print(\" \".join(cmd))\n",
    "subprocess.call(cmd)\n",
    "\n",
    "cmd = [\"bedSort\",chess31_chm13_gff_fname.rstrip(\".gff\")+\".bedPlus\",chess31_chm13_gff_fname.rstrip(\".gff\")+\".srt.bedPlus\"]\n",
    "print(\" \".join(cmd))\n",
    "subprocess.call(cmd)\n",
    "\n",
    "os.rename(chess31_chm13_gff_fname.rstrip(\".gff\")+\".srt.bedPlus\",chess31_chm13_gff_fname.rstrip(\".gff\")+\".bedPlus\")\n",
    "\n",
    "cmd = [\"bedToBigBed\",\"-type=bed12+8\",\"-tab\",\"-as=bigGenePred.as\",chess31_chm13_gff_fname.rstrip(\".gff\")+\".bedPlus\",chess3_data_dir+\"CHM13.len\",chess31_chm13_gff_fname.rstrip(\".gff\")+\".bb\",\"-extraIndex=name\"]\n",
    "print(\" \".join(cmd))\n",
    "subprocess.call(cmd)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eaa9ad88-a421-41bc-971e-e66687881567",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b7d5bfcc-f9b7-44a4-88ef-f2b832944842",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "454ce437-17f4-44c9-ab46-762503caed4d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "fe09a365-7a43-406c-8483-b45996207f29",
   "metadata": {},
   "outputs": [],
   "source": [
    "# create release\n",
    "rdir = base_dir+\"chess3.0.1/\"\n",
    "if not os.path.exists(rdir):\n",
    "    os.makedirs(rdir)\n",
    "    \n",
    "os.rename(chess31_gtf_fname,rdir+\"chess3.0.1.gtf\")\n",
    "os.rename(chess31_gff_fname,rdir+\"chess3.0.1.gff\")\n",
    "os.rename(chess31_gff_fname.rstrip(\".gff\")+\".bb\",rdir+\"chess3.0.1.bb\")\n",
    "\n",
    "\n",
    "os.rename(chess31_chm13_gtf_fname,rdir+\"chess3.0.1.CHM13.gtf\")\n",
    "os.rename(chess31_chm13_gff_fname,rdir+\"chess3.0.1.CHM13.gff\")\n",
    "os.rename(chess31_chm13_gff_fname.rstrip(\".gff\")+\".bb\",rdir+\"chess3.0.1.CHM13.bb\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "c07afc3d-b458-4030-8465-d4ca2b8c64ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "# generate files with no alts\n",
    "with open(rdir+\"chess3.0.1.primary.gtf\",\"w+\") as outFP:\n",
    "    with open(rdir+\"chess3.0.1.gtf\",\"r\") as inFP:\n",
    "        for line in inFP:\n",
    "            lcs = line.split(\"\\t\")\n",
    "            if lcs[0]==\"#\":\n",
    "                outFP.write(line)\n",
    "            if not len(lcs) == 9:\n",
    "                continue\n",
    "                \n",
    "            if \"_alt\"  in lcs[0]:\n",
    "                continue\n",
    "            \n",
    "            outFP.write(line)\n",
    "            \n",
    "            \n",
    "            \n",
    "with open(rdir+\"chess3.0.1.primary.gff\",\"w+\") as outFP:\n",
    "    with open(rdir+\"chess3.0.1.gff\",\"r\") as inFP:\n",
    "        for line in inFP:\n",
    "            lcs = line.split(\"\\t\")\n",
    "            if lcs[0]==\"#\":\n",
    "                outFP.write(line)\n",
    "            if not len(lcs) == 9:\n",
    "                continue\n",
    "                \n",
    "            if \"_alt\"  in lcs[0]:\n",
    "                continue\n",
    "            \n",
    "            outFP.write(line)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b4f0bfe2-4687-4779-9783-28b37eb0937f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# generate protein file\n",
    "cmd = [\"gffread\",\"-g\",\"hg38_p12_ucsc.fa\",\"-y\",rdir+\"chess3.0.1.protein.fa\",rdir+\"chess3.0.1.primary.gtf\"]\n",
    "print(\" \".join(cmd))\n",
    "subprocess.call(cmd)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8468717b-4789-48f7-911c-44b0a5e0f7df",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "009a967d-7204-40a9-9baa-f006eb393fce",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Generate a version of MANE with chess identifiers:\n",
    "# due to the polycistronic genes, we need a better method of linking mane IDs to chess IDs...\n",
    "chess3_attrs = definitions.get_attribute(chess3_gtf_fname,\"db_xref\",[1])\n",
    "chess3_attrs.columns = [\"tid\",\"source\",\"db_xref\"]\n",
    "chess3_attrs.head()\n",
    "\n",
    "\n",
    "c3_mane = chess3_attrs[chess3_attrs[\"source\"]==\"MANE\"].reset_index(drop=True)\n",
    "# extract refseq and gencode identifiers from the db_xref field\n",
    "c3_mane[\"rtid\"] = c3_mane[\"db_xref\"].str.split(\"RefSeq:\",n=1,expand=True)[1].str.split(\",\",n=1,expand=True)[0]\n",
    "c3_mane[\"gtid\"] = c3_mane[\"db_xref\"].str.split(\"GENCODE:\",n=1,expand=True)[1].str.split(\",\",n=1,expand=True)[0]\n",
    "mane_ctids = set(c3_mane[\"tid\"])\n",
    "print(len(mane_ctids))\n",
    "c3_mane.head()\n",
    "\n",
    "\n",
    "# now make sure all rtids and gtids are in the actual mane files\n",
    "mane_df = definitions.get_attribute(mane_refseq_gtf_fname,\"gene_id\")\n",
    "mane_df[\"tid\"] = mane_df[\"tid\"].str.split(\"-\",n=1,expand=True)[1]\n",
    "mane_rtids = set(mane_df[\"tid\"])\n",
    "print(len(mane_rtids))\n",
    "\n",
    "assert mane_rtids==set(c3_mane[\"rtid\"]),\"mismatch in tids between chess mane and mane\"\n",
    "\n",
    "\n",
    "# load map of chess ids to mane ids\n",
    "r2c = pd.Series(c3_mane.tid.values,index=c3_mane.rtid).to_dict()\n",
    "r2db_xref = pd.Series(c3_mane.db_xref.values,index=c3_mane.rtid).to_dict()\n",
    "\n",
    "\n",
    "\n",
    "# now we can write the mane file with chess transcript and gene ids\n",
    "# we shall also keep the refseq and gencode identifiers in the db_xref attribute\n",
    "with open(mane_chess_gtf_fname,\"w+\") as outFP:\n",
    "    with open(mane_refseq_gtf_fname,\"r\") as inFP:\n",
    "        for line in inFP:\n",
    "            if line[0] == \"#\":\n",
    "                outFP.write(line)\n",
    "            lcs = line.strip().split('\\t')\n",
    "            tid = lcs[8].split(\"transcript_id \\\"\", 1)[1].split(\"\\\"\", 1)[0].split(\"-\",1)[1]\n",
    "            assert tid in c2r,\"invalid tid: \"+tid\n",
    "            \n",
    "            attrs = definitions.extract_attributes(lcs[8])\n",
    "            attrs[\"db_xref\"] = r2db_xref[tid]\n",
    "            attrs[\"transcript_id\"] = r2c[tid]\n",
    "            attrs[\"gene_id\"] = r2c[tid].rsplit(\".\",1)[0]\n",
    "            \n",
    "            attr_str = definitions.to_attribute_string(attrs)\n",
    "            \n",
    "            lcs[8] = attr_str\n",
    "            \n",
    "            outFP.write(\"\\t\".join(lcs)+\"\\n\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
